lib/puppet/indirector/catalog/static_compiler.rb

```diff
@@ -188,9 +189,16 @@ def get_child_resources(request, catalog, resource, file)
   # @param children [Array<Puppet::Resource>]
   # @param catalog [Puppet::Resource::Catalog]
   def remove_existing_resources(children, catalog)
-    existing_names = catalog.resources.collect { |r| r.to_s }
-    both = (existing_names & children.keys).inject({}) { |hash, name| hash[name] = true; hash }
-    both.each { |name| children.delete(name) }
+    # build a list of stringified file resources from the children
+    child_names = children.collect { |k, r| r.to_s }
+    # build a list of stringified file resources from the catalog
+    existing_names = catalog.resources.inject([]) do |list, res|
+      list << res.to_s if res.type == 'File'
+      list
+    end
```

Comment:
Rather than iterate over every resource in the catalog (which could be large) and built up an array of existing file resource names, we could iterate over the recursed children, and check to see if it's already in the catalog, something like:

```
    relative_paths = children.keys
    relative_paths.each do |relative_path|
      child = children[relative_path]
      if catalog.resource(child.ref)
        Puppet.debug("Resource #{child.ref} already managed, removing from recursed children")
        children.delete(relative_path)
      end
    end
```

Ideally we could just iterate over `children`, but I don't think it's safe to mutate the Array while it's being iterated over.